### DISCUSSION TITLE
Error mapping one-to-many self-relations

### DISCUSSION THREAD

**Discussion by spidertwin2** [prisma/prisma#7897] [Q&A] [answered]
Salut!

For each `Post`, the user can write `comments`.  Each `Comment`, can have multiple `replies`.  I am unable to correctly map this in my schema.

`Error validating field 'replies' in model 'Comment' : The relation field 'replies' on Model 'Comment' is missing an opposite relation field on the model 'Comment'. Either run 'prisma format' or add it manually.`

What should the opposite relation field in `Comment` be?

```prisma
model Post {
  id        String   @id @default(cuid())
  title     String?
  body      String?

  comments Comment[] @relation("CommentsOnPost")
  author   User?     @relation("WrittenPosts", fields: [authorId], references: [id])
  authorId Int
}

model Comment {
  id        String   @id @default(cuid())
  body      String

  replies  Comment[] @relation("RepliesOnComment")  //error here
  post     Post      @relation("CommentsOnPost", fields: [postId], references: [id])
  postId   String
  author   User      @relation("WrittenComments", fields: [authorId], references: [id])
  authorId Int
}

```

Just wondering, doesn't this create an infinite loop problem where each reply can be replied to?  Is this bad? Is there a way to solve for this?

Thank you! :watermelon:

**Comment by ryands17** [comment, marked as answer]
@spidertwin2 👋 

You are missing the extra relation that is required for self-relations as this is a 1-many self relation. So your schema should look like:

```prisma
model User {
  id       Int       @id @default(autoincrement())
  name     String
  posts    Post[]    @relation("WrittenPosts")
  comments Comment[] @relation("WrittenComments")
}

model Post {
  id    String  @id @default(cuid())
  title String?
  body  String?

  comments Comment[] @relation("CommentsOnPost")
  author   User?     @relation("WrittenPosts", fields: [authorId], references: [id])
  authorId Int
}

model Comment {
  id   String @id @default(cuid())
  body String

  replies        Comment[] @relation("RepliesOnComment")
  post           Post      @relation("CommentsOnPost", fields: [postId], references: [id])
  postId         String
  author         User      @relation("WrittenComments", fields: [authorId], references: [id])
  authorId       Int
  replyOnComment Comment?  @relation("RepliesOnComment", fields: [commentId], references: [id])
  commentId      String?
}
```

If you're using VSCode with the [Prisma plugin](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) and you save the above schema, it will automatically add the relation for you. You can read more about self relations [here](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#one-to-many-self-relations).

**Reply by spidertwin2** [reply]
Thanks @ryands17!

What do `replyOnComment` and `commentId` really do?  

This relation expresses the following: 

- "a comment has zero or one replies"
- "a comment can have zero or more replies"

It seems a bit redundant, doesn't the second include the first?

**Reply by ryands17** [reply]
Prisma always has virtual relations for each foreign key, so yes that is redundant and you do not need to use that. You would just interact with `replies` as `replyOnComment` and `commentId` are just references for the 1-many relation. 

For e.g. Think of a `User` and `Post` model where 1 user can have many posts:

```prisma
model User {
  id     Int     @id @default(autoincrement())
  posts  Post[]
}

model Post {
  id        Int   @id @default(autoincrement())
  user    User  @relation(fields: [userId], references: [id])
  userId  Int   // relation scalar field  (used in the `@relation` attribute above)
}
```

When you create `posts`, a `userId` is automatically created along with the virtual field `user` in the `Post` model. This concept is the same except this time, it is all in the same table due to the relation being a self-relation.

Let me know if that helps :)

**Reply by spidertwin2** [reply]
makes more sense now, merci!

:cake:

**Reply by ErickRuano** [reply]
> @spidertwin2 👋
> 
> You are missing the extra relation that is required for self-relations as this is a 1-many self relation. So your schema should look like:
> 
> ```prisma
> model User {
>   id       Int       @id @default(autoincrement())
>   name     String
>   posts    Post[]    @relation("WrittenPosts")
>   comments Comment[] @relation("WrittenComments")
> }
> 
> model Post {
>   id    String  @id @default(cuid())
>   title String?
>   body  String?
> 
>   comments Comment[] @relation("CommentsOnPost")
>   author   User?     @relation("WrittenPosts", fields: [authorId], references: [id])
>   authorId Int
> }
> 
> model Comment {
>   id   String @id @default(cuid())
>   body String
> 
>   replies        Comment[] @relation("RepliesOnComment")
>   post           Post      @relation("CommentsOnPost", fields: [postId], references: [id])
>   postId         String
>   author         User      @relation("WrittenComments", fields: [authorId], references: [id])
>   authorId       Int
>   replyOnComment Comment?  @relation("RepliesOnComment", fields: [commentId], references: [id])
>   commentId      String?
> }
> ```
> 
> If you're using VSCode with the [Prisma plugin](https://marketplace.visualstudio.com/items?itemName=Prisma.prisma) and you save the above schema, it will automatically add the relation for you. You can read more about self relations [here](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#one-to-many-self-relations).

I am following this exact schema structure and although it mostly works, I am not being able to fetch nested records with:

```json
{
  "include": {
    "children" : true
  } 
}
```

This is the schema I'm working with:

```prisma
model Folder {
  id            Int      @id @default(autoincrement())
  name          String
  folderId Int?
  parentFolder   Folder?     @relation("FoldersOnFolder", fields: [folderId], references: [id])
  children  Folder[]        @relation("FoldersOnFolder")
}

```

Any idea on how to get it to work would be much appreciated.

**Reply by ryands17** [reply]
Is there any error that you're getting or are the records not available? The schema seems correct to me so could you check if you have added the `children` correctly via `create` or `connect`?

**Reply by ErickRuano** [reply]
Thanks for the quick support. I thought I wasn't implementing the relationships properly but once you helped me validate it was correct I found the issue was actually on my side.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
